src/libopensc/card-openpgp.c

```diff
@@ -2735,54 +2934,73 @@ pgp_build_extended_header_list(sc_card_t *card, sc_cardctl_openpgp_keystore_info
 		"modulus"
 	};
 	size_t comp_to_add = 3;
-	size_t req_e_len = 0;     /* The exponent length specified in Algorithm Attributes */
-	pgp_blob_t *alat_blob;
+	/* The maximum exponent length is 32 bit, as set on card
+	 * we use this variable to check against actual exponent_len */
+	size_t max_e_len = 0x20>>3;
```

Comment:
Magic number; maybe better add a symbolic constant in cardctl.h

Reply:
I am not sure. We only need this value in this function and in [pgp_upgate_new_algo_attr](https://github.com/OpenSC/OpenSC/blob/fc7dba0a40c69e32cb8a0a199a1773f236f44f95/src/libopensc/card-openpgp.c#L2295-L2299). I'd prefer to have a clear comment in these parts, instead of having a constant in cardctl.h.

Reply:
I'd still go for the symbolic constant.
The magic 32 for the exponent length is also used in line 2299 of card-openpgp.c:
```
                /* OpenPGP Card only accepts 32bit as exponent length field,
                 * although you can import keys with smaller exponent;
                 * thus we don't change rsa.exponent_len, but ignore it here */
                data[3] = 0x00;
                data[4] = 0x20;
                data[5] = 0x00; /* Import-Format of private key (e,p,q) */
```
And there is something related around line 3119
```
                /* we only support exponent of maximum 32 bits */
                if (key_info->u.rsa.e_len > 4) {
                        sc_log(card->ctx,
                                   "Exponent %"SC_FORMAT_LEN_SIZE_T"u-bit (>32) is not supported.",
                                   key_info->u.rsa.e_len * 8);
                        LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
                }
```
BTW, is it correct that `sc_cardctl_openpgp_keygen_info_t`'s `exponent_len` is measured in bits, while `sc_cardctl_openpgp_keystore_info_t`'s `e_len` is in bytes?

Reply:
Okay, I'll change it to a constant.

> BTW, is it correct that sc_cardctl_openpgp_keygen_info_t's exponent_len is measured in bits, while sc_cardctl_openpgp_keystore_info_t's e_len is in bytes?

Good point. As far as I can see, this was done, because sc_pkcs15_prkey_t does use bytes. This value is copied to e_len in openpgp_store_key function of pkcs15-openpgp.c. It does make much more sense to transform it to bits there and thus use e_len consistent to exponent_len of sc_cardctl_openpgp_keygen_info_t. This saves us all the shifting operations in card-openpgp.c.

I'll change this these days.